temporal-sdk/src/main/java/io/temporal/internal/common/WorkflowExecutionUtils.java

```diff
@@ -256,11 +256,10 @@ public static HistoryEvent getInstanceCloseEvent(
                       .withDeadline(expirationInRetry)
                       .getWorkflowExecutionHistory(r);
                 });
-      }
-      if (response == null || !response.hasHistory()) {
-        continue;
-      }
-      if (timeout != 0 && System.currentTimeMillis() - start > unit.toMillis(timeout)) {
+        if (response == null || !response.hasHistory()) {
```

Comment:
Note that there is a slight behavioral change here, as before we've used to check current time AFTER the call to the server, now we are checking that condition BEFORE. This means that code below will be executed one more time comparing to what it was, and it is possible that we return an event for a workflow that timed out already. I think this is not great, and I would prefer to always check the timeout condition with current time after the service call.

Reply:
Yeah, I actually made this change intentionally.
It looks pointless to wait for an event, actually get it and verify that it's a real event only to timeout after it. 
I personally think that it's actually a fix for the edge case.

Why do you think it's so critical to throw timeout even if we already spent this time and we already got the event?

Reply:
I think it's okay given that it's not a workflow timeout, just wanted to point out that there is a change in behavior.

Reply:
It's really a super tiny edge case here.
We enforce and actively work with timeout in GrpcRetryer here:
```
GrpcRetryer.retryWithResult(
                retryOptions,
                () -> {
                  long elapsedInRetry = System.currentTimeMillis() - start;
                  Deadline expirationInRetry =
                      Deadline.after(
                          unit.toMillis(timeout) - elapsedInRetry, TimeUnit.MILLISECONDS);
                 ... 
)
```
So, it's possible that we actually met the deadline, got the event from the Retryer and after that discarded because one 1 more millisecond ticked or something.
So, it's not like this change really makes it possible to significantly overdue the timeout, it's still fully enforced in the Retryer and in the lambda that is passed to the Retryer.

Reply:
Yep, let's keep it like it is then.